test/jasmine/tests/axes_test.js

```diff
@@ -613,6 +613,204 @@ describe('Test axes', function() {
+            .then(function() {
+                assertRangeDomain('xaxis', [-3.5, 4.5], [0, 1], [0, 1]);
+                assertRangeDomain('yaxis', [-0.5, 1.5], [0, 1], [0.5, 1]);
+
+                return Plotly.relayout(gd, {'xaxis.range': [0, 1]});
+            })
+            .then(function() {
+                assertRangeDomain('xaxis', [0, 1], [0, 1], [0, 0.25]);
+                assertRangeDomain('yaxis', [-0.5, 1.5], [0, 1], [0, 1]);
+            })
+            .catch(failTest)
+            .then(done);
+        });
+
+        it('autoranges consistently with padding', function(done) {
+            var xAutoPad = 0.09523809523809526;
+            var xAutorange = [-xAutoPad, 1 + xAutoPad];
+            var yAutoPad = 0.15476190476190477;
+            var yAutorange = [-yAutoPad, 1 + yAutoPad];
+            Plotly.plot(gd, [
+                {y: [0, 1], mode: 'markers', marker: {size: 4}},
+                {y: [0, 1], mode: 'markers', marker: {size: 4}, xaxis: 'x2', yaxis: 'y2'}
+            ], {
+                xaxis: {domain: [0, 0.5], constrain: 'domain'},
+                yaxis: {constrain: 'domain', scaleanchor: 'x'},
+                xaxis2: {domain: [0.5, 1], constrain: 'domain'},
+                yaxis2: {constrain: 'domain', scaleanchor: 'x2'},
+                // plot area 200x200px, so y axes should be squished to
+                // (a little over due to autoranging) half their input domain
+                width: 400,
+                height: 400,
+                margin: {l: 100, r: 100, t: 100, b: 100, p: 0},
+                showlegend: false
+            })
+            .then(function() {
+                assertRangeDomain('xaxis', xAutorange, [0, 0.5], [0, 0.5]);
+                assertRangeDomain('yaxis', yAutorange, [0, 1], [0.225, 0.775]);
+                assertRangeDomain('xaxis2', xAutorange, [0.5, 1], [0.5, 1]);
+                assertRangeDomain('yaxis2', yAutorange, [0, 1], [0.225, 0.775]);
+
+                return Plotly.relayout(gd, {'xaxis.range': [-1, 2]});
+            })
+            .then(function() {
+                assertRangeDomain('xaxis', [-1, 2], [0, 0.5], [0, 0.5]);
+                assertRangeDomain('yaxis', [-0.39, 1.39], [0, 1], [0.3516667, 0.6483333]);
+                assertRangeDomain('xaxis2', xAutorange, [0.5, 1], [0.5, 1]);
+                assertRangeDomain('yaxis2', yAutorange, [0, 1], [0.225, 0.775]);
+
+                return Plotly.relayout(gd, {'xaxis.autorange': true});
+            })
+            .then(function() {
+                assertRangeDomain('xaxis', xAutorange, [0, 0.5], [0, 0.5]);
+                assertRangeDomain('yaxis', yAutorange, [0, 1], [0.225, 0.775]);
+                assertRangeDomain('xaxis2', xAutorange, [0.5, 1], [0.5, 1]);
+                assertRangeDomain('yaxis2', yAutorange, [0, 1], [0.225, 0.775]);
+            })
+            .catch(failTest)
+            .then(done);
+        });
+
+        it('can constrain date axes', function(done) {
```

Comment:
category and log axis domain and range constraint tests to match the category one -> https://github.com/plotly/plotly.js/pull/1767/commits/67ea3a8344e8907ebc53c035d8e1222954fd1f30